pandas/tests/io/parser/test_common.py

```diff
@@ -2253,3 +2262,71 @@ def test_dict_keys_as_names(all_parsers):
     result = parser.read_csv(StringIO(data), names=keys)
     expected = DataFrame({"a": [1], "b": [2]})
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("io_class", [StringIO, BytesIO])
+@pytest.mark.parametrize("encoding", [None, "utf-8"])
+def test_read_csv_file_handle(all_parsers, io_class, encoding):
+    """
+    Test whether read_csv does not close user-provided file handles.
+
+    GH 36980
+    """
+    parser = all_parsers
+    expected = DataFrame({"a": [1], "b": [2]})
+
+    content = "a,b\n1,2"
+    if io_class == BytesIO:
+        content = content.encode("utf-8")
+    handle = io_class(content)
+
+    tm.assert_frame_equal(parser.read_csv(handle, encoding=encoding), expected)
+    assert not handle.closed
+
+
+def test_memory_map_compression_error(c_parser_only):
+    """
+    c-parsers do not support memory_map=True with compression.
+
+    GH 36997
+    """
+    parser = c_parser_only
+    df = DataFrame({"a": [1], "b": [2]})
+    msg = (
+        "read_csv does not support compression with memory_map=True. "
+        + "Falling back to memory_map=False. Please use "
+        + "memory_map=False instead."
+    )
+
+    with tm.ensure_clean() as path:
+        df.to_csv(path, compression="gzip", index=False)
+
+        with pytest.raises(ValueError, match=msg):
+            parser.read_csv(path, memory_map=True, compression="gzip")
+
+
+def test_memory_map_file_handle_error(c_parser_only):
+    """
+    c-parsers do support only string-like files with memory_map=True.
+
+    GH 36997
+    """
+    parser = c_parser_only
+    expected = DataFrame({"a": [1], "b": [2]})
+    msg = (
+        "read_csv does support only string-like objects with "
+        + "memory_map=True and engine='c'. Falling back to "
+        + "memory_map=False. Please use engine='python' instead."
+    )
+
+    handle = StringIO("a,b\n1,2")
+
+    with pytest.raises(ValueError, match=msg):
+        parser.read_csv(handle, memory_map=True)
+
+    # verify that python engine supports the same call
+    handle.seek(0)
+    tm.assert_frame_equal(
+        read_csv(handle, memory_map=True, engine="python"),
+        expected,
+    )
```

Comment:
This ends up testing the Python parser twice.

We should just use the `all_parsers` fixture and check the engine type.

Reply:
done